docs/NORMALIZATION.md

```diff
@@ -0,0 +1,34 @@
+# Address Normalization
+
+Cordial Systems APIs use blockchain addresses as identifiers for both addresses and assets, for example
+
+- `chains/ETH/addresses/0x4e27f5efe252fb7fbee02f503be881ac2d20f679` is the name of an address on Ethereum, and
+- `chains/ETH/assets/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48` is the name of an asset on Ethereum (the USDC stablecoin)
+
+Two steps of normalization are in play:
+
+- non-destructive / canonicalization: Some chains use multiple equivalent formats for an address, for instance Ethereum uses 0x-prefixed lower-case hex, but also a mixed-case variant encoding a checksum
+- destructive / irreversible: As a general principle, in Cordial APIs one can "get" a resource by sending an HTTP GET request corresponding to the resource name. Hence we restrict names to contain characters that are valid in a URL, and more specifically, to alphanumeric (`[0-9A-Za-z]`), dash (`-`) and underscore (`_`).
+
+The first step is chain-specific, `crosschain` defines and uses a single canonical format for each address on each chain. The intent of this document is to specify this canonical format, for all chains supported by `crosschain`.
+
+The second step is generic, we replace whitespace with underscore (`_`) and all other invalid characters with dashes (`-`). Since this is not reversible, an address resource will have a field `address` containing the original (canonical) address, and a (token) asset resource will have a field `contract` containing the original (canonical) address.
```

Comment:
I would add that this second part is not done in this library.  It's done only for testing to ensure compatibility with external APIs / primary keys.